dmoj/executors/filesystem_policies.py

```diff
@@ -0,0 +1,109 @@
+from typing import List, Union
+
+
+class ACCESS_MODE:
+    NONE = 0
+    EXACT = 1
+    RECURSIVE = 2
+
+
+class Dir:
+    def __init__(self):
+        self.access_mode = ACCESS_MODE.NONE
+
+        self.map = {}
+
+
+class File:
+    pass
+
+
+class ExactFile:
+    def __init__(self, path : str):
+        self.path = path
+
+
+class ExactDir:
+    access_mode = ACCESS_MODE.EXACT
+
+    def __init__(self, path : str):
+        self.path = path
+
+
+class RecursiveDir:
+    access_mode = ACCESS_MODE.RECURSIVE
+
+    def __init__(self, path : str):
+        self.path = path
+
+
+Rule = Union[ExactFile, ExactDir, RecursiveDir]
+
+
+class FilesystemPolicy:
+    def __init__(self, rules : List[Rule]):
+        self.rules = rules
+
+        self.built = False
+
+    def __add__(self, other):
+        assert not self.built and not other.built, "Refusing to merge built policies: merge before building"
+
+        return FilesystemPolicy(self.rules + other.rules)
+
+    def build(self):
+        self.root = Dir()
+
+        for rule in self.rules:
+            self.add_rule(rule)
+
+        self.built = True
+
+    def add_rule(self, rule : Rule):
+        node = self.root
+
+        path_split = rule.path.split("/")
+
+        for index in range(len(path_split) - 1):
```

Comment:
We don't ever use `index` besides for indexing into `path_split`. Would it be more appropriate to iterate over the list directly? i.e. `for part in path_split[:-1]:`

(Similar comment for `check`).

Reply:
Maybe. This would require making a copy of the list for `add_rule`, which would impact speed. This is probably fine though. In `check`, there wouldn't be a performance hit, which is important.